File: go/builtins/funcTrimEnd.go
package builtins

import (
	"strings"
	"unicode"

	"github.com/jplorg/jpl/go/jpl"
	"github.com/jplorg/jpl/go/library"
)

var funcTrimEnd = enclose(func(runtime jpl.JPLRuntime, signal jpl.JPLRuntimeSignal, next jpl.JPLPiper, input any, args ...any) ([]any, error) {
	t, err := library.Type(input)
	if err != nil {
		return nil, err
	}

	switch t {
	case jpl.JPLT_STRING:
		alteredValue, err := library.AlterValue(input, jpl.JPLModifierFunc(func(value any) (any, jpl.JPLError) {
			return strings.TrimRightFunc(value.(string), unicode.IsSpace), nil
		}))
		if err != nil {
			return nil, err
		}
		return next.Pipe(alteredValue)
	}

	return nil, nil
})

File: js/lib/builtins/funcTrimEnd.ts
import { JPLTypeError, type JPLRuntimeSignal } from '@/library';
import type JPLRuntime from '@/runtime';

export default async function builtin(
  runtime: JPLRuntime,
  _signal: JPLRuntimeSignal,
  next: (output: unknown) => Promise<unknown[]> | unknown[],
  input: unknown,
): Promise<unknown[]> {
  const t = runtime.type(input);

  switch (t) {
    case 'string':
      return next(
        await runtime.alterValue(input, (value) => (value as string).trimEnd()),
      );

    default:
  }

  throw new JPLTypeError(
    '%s (%*<100v) cannot be trimmed',
    t,
    runtime.unwrapValue(input),
  );
}
